trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java

```diff
@@ -1353,6 +1430,84 @@ private Result handleAtomicExchangeN(final IDispatcher main, final IASTFunctionC
+		final ExpressionResult successMemoryOrder =
+				mExprResultTransformer.transformDispatchSwitchRexBoolToInt(main, loc, arguments[4]);
+		final ExpressionResult failureMemoryOrder =
+				mExprResultTransformer.transformDispatchSwitchRexBoolToInt(main, loc, arguments[5]);
+		final var resultBuilder = new ExpressionResultBuilder().addAllExceptLrValue(pointer, expectedResult,
+				desiredResult, weakResult, successMemoryOrder, failureMemoryOrder);
+
+		final var boolType = new CPrimitive(CPrimitives.BOOL);
+		final var success = mAuxVarInfoBuilder.constructAuxVarInfo(loc, boolType, AUXVAR.RETURNED);
+		final var pointerRead = mExprResultTransformer.readPointerValue(loc, pointer.getLrValue());
+		final var expectedRead = mExprResultTransformer.readPointerValue(loc, expectedResult.getLrValue());
+		final var pointerWrite = mExprResultTransformer.makePointerAssignment(loc, pointer.getLrValue(),
+				desiredResult.getLrValue().getValue());
+
+		final var successBoolean =
+				mExpressionTranslation.constructBinaryEqualityExpression(loc, IASTBinaryExpression.op_notequals,
+						success.getExp(), boolType, mExpressionTranslation.constructZero(loc, boolType), boolType);
+		final var atomicBody = new ExpressionResultBuilder().addAuxVarWithDeclaration(success)
+				.addStatement(StatementFactory.constructIfStatement(loc, weakResult.getLrValue().getValue(),
+						new Statement[] { new HavocStatement(loc, new VariableLHS[] { success.getLhs() }) },
+						new Statement[] { StatementFactory.constructAssignmentStatement(loc,
+								new LeftHandSide[] { success.getLhs() },
+								new Expression[] { mExpressionTranslation.constructLiteralForIntegerType(loc, boolType,
+										BigInteger.ONE) }) }))
+				.addAllExceptLrValue(pointerRead, expectedRead).addAllExceptLrValueAndStatements(pointerWrite)
+				.addStatement(StatementFactory.constructIfStatement(loc,
+						new BinaryExpression(loc, Operator.LOGICAND, successBoolean,
+								new BinaryExpression(loc, Operator.COMPEQ, pointerRead.getLrValue().getValue(),
+										expectedRead.getLrValue().getValue())),
+						pointerWrite.getStatements().toArray(Statement[]::new),
+						new Statement[] { StatementFactory.constructAssignmentStatement(loc,
+								new LeftHandSide[] { success.getLhs() }, new Expression[] { mExpressionTranslation
+										.constructLiteralForIntegerType(loc, boolType, BigInteger.ZERO) }) }))
+				.build();
+		final var atomic = applyMemoryOrder(loc, atomicBody, successMemoryOrder.getLrValue().getValue(),
+				failureMemoryOrder.getLrValue().getValue());
+
+		final var expectedWrite = mExprResultTransformer.makePointerAssignment(loc, expectedResult.getLrValue(),
+				pointerRead.getLrValue().getValue());
+		return resultBuilder.addAllExceptLrValue(atomic).addAllExceptLrValueAndStatements(expectedWrite)
+				.addStatement(StatementFactory.constructIfStatement(loc, successBoolean, new Statement[0],

```

Comment:
Minor note: Is there any reason why you create `if (success) {} else st` instead of `if (!success) st` here? It should not make a difference on the CFG, but the Boogie code of the latter looks slightly better?